lib/Sema/BuilderTransform.cpp

```diff
@@ -613,3 +588,80 @@ ConstraintSystem::TypeMatchResult ConstraintSystem::applyFunctionBuilder(
                 locator);
   return getTypeMatchSuccess();
 }
+
+namespace {
+
+/// Pre-check all the expressions in the closure body.
+class PreCheckFunctionBuilderClosure : public ASTWalker {
+  TypeChecker &TC;
+  ClosureExpr *Closure;
+  bool HasReturnStmt = false;
+  bool HasError = false;
+public:
+  PreCheckFunctionBuilderClosure(TypeChecker &tc, ClosureExpr *closure)
+    : TC(tc), Closure(closure) {}
+
+  FunctionBuilderClosurePreCheck run() {
+    Stmt *oldBody = Closure->getBody();
+
+    Stmt *newBody = oldBody->walk(*this);
+
+    // If the walk was aborted, it was because we had a problem of some kind.
+    assert((newBody == nullptr) == (HasError || HasReturnStmt) &&
+           "unexpected short-circuit while walking closure body");
+    if (!newBody) {
+      if (HasError)
+        return FunctionBuilderClosurePreCheck::Error;
+
+      return FunctionBuilderClosurePreCheck::HasReturnStmt;
+    }
+
+    assert(oldBody == newBody && "pre-check walk wasn't in-place?");
+
+    return FunctionBuilderClosurePreCheck::Okay;
+  }
+
+  virtual std::pair<bool, Expr *> walkToExprPre(Expr *E) {
```

Comment:
Nit: did you mean `override` instead of `virtual` here and in `walkToStmtPre`?

Reply:
Yes, I'll fix that.  Thanks!